### ISSUE TITLE
Benchmarking/Profiling

### ISSUE THREAD

**Issue by austinlparker** [open-telemetry/opentelemetry-demo#1601]
We could stand to set up a continuous benchmark/profile with the demo. This would involve (to start) a version with no instrumentation and one as stands. We would run these tests after new releases of otel dependencies and publish the results.

**Comment by rogercoll** [comment]
[Just recently](https://opentelemetry.io/blog/2024/elastic-contributes-continuous-profiling-agent/), Elastic contributed to OpenTelemetry with its continuous profiling agent. Once the agent is available, I think it could be integrated in the demo to tackle this issue. What are your thoughts on a universal profiler for all the services?

**Comment by austinlparker** [comment]
yeah, i was thinking we could use the profiler for this.

**Comment by austinlparker** [comment]
really my motivation here is mostly to provide some sort of starting point/baseline for discussing/thinking about performance and overhead of the sdk and various instrumentations.

**Comment by janssenlima** [comment]
Guys. Any updates on this?

**Comment by julianocosta89** [comment]
@janssenlima, tudo bem? 🇧‍🇷 😊 

We are more than happy to accept contributions on this.
AFAIK there is no OSS tool that is currently receiving OTLP profiles.

During KubeCon EU Damien and Felix shared this: https://www.youtube.com/watch?v=aXGP2YLeG_A
But TBH I'm not sure Profiles are stable enough to be deployed and maintained.

**Comment by florianl** [comment]
When talking to people at KubeCon Amsterdam 2026, having OTel Profiling as part of the OTel demo was a highly requested feature.

FYI: @osullivandonal

**Comment by julianocosta89** [comment]
@florianl and @osullivandonal, yes!
How difficult would it be to have a devfiler container configured and running?

**Comment by osullivandonal** [comment]
Just seeing this now, this is something we should try and get into the demo. I'll have a look at this and get back here. cc. @florianl @julianocosta89

**Comment by florianl** [comment]
> How difficult would it be to have a devfilter container configured and running?

devfiler is a desktop application with no real storage or backend. It is focused on fast local development and it is not maintained. So it is not suitable for something like the OTel Demo, I think.

**Comment by julianocosta89** [comment]
@florianl the Demo don't need to be production ready.
Actually currently we have limits for Jaeger and Prometheus and the data is purged after x amount of data is ingested. We could reproduce the same on the Profile side.
I'd love if we could have any vendor agnostic profile visualization tool on the Demo

**Comment by florianl** [comment]
As mentioned, devfiler is a desktop application with wayland and x11 rendering. it is not suited for a web focused deployment. I think, there will be no visualization stack deployed in the OTel demo environment, to make sure x11 or wayland are available.

**Comment by osullivandonal** [comment]
I have a PR open than demonstrates how we can add profiling to the OTel-Demo using Pyroscope as the backend.

<img width="2455" height="1196" alt="Image" src="https://github.com/user-attachments/assets/b6b64a07-729f-41c0-bd49-62eb818b416b" />

## Demonstration Workflow

To demonstrate this I can successfully view hot code paths, for example if we turn on the feature flag for high cpu for the ad service:

<img width="2498" height="1270" alt="Image" src="https://github.com/user-attachments/assets/17ee704e-ad79-4532-b745-1e88c08a598f" />

It would be great to get OTel profiling maintainers view of this, cc. @florianl 

Unfortunately this is not vendor agnostic as this is a Grafana labs product. Pyroscope was chosen here to its ability to support OTLP.

Another option we could look at is [Parca](https://www.parca.dev/), but I am not sure how much OTel support they have.

**Comment by christos68k** [comment]
> I have a PR open than demonstrates how we can add profiling to the OTel-Demo using Pyroscope as the backend.
> Unfortunately this is not vendor agnostic as this is a Grafana labs product. Pyroscope was chosen here to its ability to support OTLP.
> 
> Another option we could look at is [Parca](https://www.parca.dev/), but I am not sure how much OTel support they have.

There's no clear solution for a backend here, as OpenTelemetry Profiles doesn't specify a backend. 

I think we should not be pushing a single vendor, as it shows bias (multiple vendors support OTLP Profiles). On the other hand, adding support for multiple vendors may be expanding the scope unnecessarily. 

One option is to start with a baseline solution that deploys Profiles receiver inside OTel Collector and converts the profiling data into metrics that are then visualized (e.g. something like https://github.com/elastic/opentelemetry-collector-components/tree/main/connector/profilingmetricsconnector or even simpler).

CC: @chrsmark

**Comment by florianl** [comment]
Hey 👋
maybe [firepit](https://github.com/florianl/firepit/tree/main) can be an option:

<img width="1918" height="996" alt="Image" src="https://github.com/user-attachments/assets/0e6987a2-879d-4457-a709-b84f044a4a2b" />

It is a project I'm using for some time now, as I'm experimenting with some other profiling types.

[firepit](https://github.com/florianl/firepit/tree/main) is a minimal, vendor-agnostic Go application designed for demoing and development. It visualizes OTel profiles as flamegraphs through a web UI on port 8080 and listens for profiles on ports 4317 (gRPC) and 4318 (HTTP). Given its specific focus, the application employs non-persistent storage and does not include authentication.

### Getting started
Run firepit:
```
$ make docker-run 
docker run -it --rm \
	-p 4317:4317 \
	-p 4318:4318 \
	-p 8080:8080 \
	firepit
2026/04/28 06:19:26 INFO Configuration loaded grpc_addr=:4317 http_addr=:4318 web_addr=:8080 profile_ttl=5m0s cleanup_interval=30s max_body_size=33554432 max_storage_bytes=524288000
2026/04/28 06:19:26 INFO OTLP HTTP server listening addr=:4318
2026/04/28 06:19:26 INFO Web UI listening addr=:8080
2026/04/28 06:19:26 INFO Open browser to url=http://localhost:8080
2026/04/28 06:19:26 INFO gRPC server listening addr=:4317
```
Let a OTel collector report profiles to firepit:
```
receivers:
  profiling:
    off_cpu_threshold: 0.1

exporters:
  otlp/firepit:
    endpoint: 127.0.0.1:4317
    tls:
      insecure: true
      insecure_skip_verify: true

service:
  pipelines:
    profiles:
      receivers: [profiling]
      exporters: [otlp/firepit]
```

**Comment by puckpuck** [comment]
I don't hate the firepit idea here. Small lightweight and vendor agnostic. 

I am concerned about future support though. Like if firepit has a security issue how does that get resolved since this is someone's personal project?

**Comment by florianl** [comment]
Yeah, [firepit](https://github.com/florianl/firepit/tree/main) is a personal project so far. But I'm also happy to donate it, if this helps. I also don't see a problem onboarding contributors. Otherwise, I'm trying to follow security best practices - e.g. govulncheck runs on every PR.

With the focus on development and demoing, [firepit](https://github.com/florianl/firepit/tree/main) tries to keep a minimal footprint and so also keeping the risk for security related issues minimal. By accepting only OTel profiles it also can be replaced any time with something else, that accepts and visualizes OTel profiles.

**Comment by julianocosta89** [comment]
@florianl We discussed firepit on the SIG meeting and everyone on the call approved the use of it!
i believe @osullivandonal will change his current PR to use firepit 🔥 

Thx for that!
